File: native/macos/InkLoopMeetingCompanion/Sources/InkLoopMeetingAdapter/Contracts.swift
import Foundation

public let meetingSessionSchemaVersion = "inkloop.meeting_session.v1"
public let meetingAudioChunkSchemaVersion = "inkloop.meeting_audio_chunk.v1"
public let meetingRealtimeAudioFrameSchemaVersion = "inkloop.meeting_realtime_audio_frame.v1"

public enum MeetingPlatform: String, Codable, Sendable {
  case googleMeet = "google_meet"
  case zoom
}

public enum MeetingStartMode: String, Codable, Sendable {
  case automatic
  case manual
}

public enum MeetingSessionStatus: String, Codable, Sendable {
  case detected
  case recording
  case paused
  case sealed
}

public enum MeetingAudioTrack: String, Codable, CaseIterable, Sendable {
  case mic
  case remote
}

public enum MeetingStopReason: String, Codable, Sendable {
  case meetingEndConfirmed = "meeting_end_confirmed"
  case manual
  case interruptedSessionRecovered = "interrupted_session_recovered"
}

public struct DetectedMeeting: Equatable, Sendable {
  public let platform: MeetingPlatform
  public let meetingReference: String
  public let detectionID: String
  public let detectedAtMonotonicMilliseconds: Int64

  public init(
    platform: MeetingPlatform,
    meetingReference: String,
    detectionID: String,
    detectedAtMonotonicMilliseconds: Int64
  ) {
    self.platform = platform
    self.meetingReference = meetingReference
    self.detectionID = detectionID
    self.detectedAtMonotonicMilliseconds = detectedAtMonotonicMilliseconds
  }
}

public struct ConfirmedMeetingEndEvidence: Codable, Equatable, Sendable {
  public let adapter: String
  public let signal: String
  public let providerEventID: String?
  public let observedAtWallClockMilliseconds: Int64?

  public init(
    validatingAdapter adapter: String,
    signal: String,
    providerEventID: String? = nil,
    observedAtWallClockMilliseconds: Int64? = nil
  ) throws {
    guard !adapter.isEmpty, !signal.isEmpty else {
      throw MeetingAdapterError.invalidConfirmedEndEvidence
    }
    guard Self.confirmedSignals.contains(signal) else {
      throw MeetingAdapterError.weakMeetingEndSignal(signal)
    }
    self.adapter = adapter
    self.signal = signal
    self.providerEventID = providerEventID
    self.observedAtWallClockMilliseconds = observedAtWallClockMilliseconds
  }

  private static let confirmedSignals: Set<String> = [
    "meeting_call_ended",
    "platform_meeting_ended",
    "user_left_meeting",
  ]

  enum CodingKeys: String, CodingKey {
    case adapter
    case signal
    case providerEventID = "provider_event_id"
    case observedAtWallClockMilliseconds = "observed_at_wall_clock_ms"
  }
}

public struct MeetingAudioChunk: Codable, Equatable, Sendable {
  public let schemaVersion: String
  public let chunkID: String
  public let sessionID: String
  public let track: MeetingAudioTrack
  public let sequence: Int
  public let startMonotonicMilliseconds: Int64
  public let endMonotonicMilliseconds: Int64
  public let checksum: String
  public let byteLength: Int
  public let sealed: Bool
  public let codec: String?
  public let sampleRateHertz: Int?
  public let channelCount: Int?

  public init(
    chunkID: String,
    sessionID: String,
    track: MeetingAudioTrack,
    sequence: Int,
    startMonotonicMilliseconds: Int64,
    endMonotonicMilliseconds: Int64,
    checksum: String,
    byteLength: Int,
    codec: String? = nil,
    sampleRateHertz: Int? = nil,
    channelCount: Int? = nil
  ) throws {
    guard !chunkID.isEmpty,
      !sessionID.isEmpty,
      sequence >= 0,
      startMonotonicMilliseconds >= 0,
      endMonotonicMilliseconds > startMonotonicMilliseconds,
      !checksum.isEmpty,
      byteLength >= 0
    else {
      throw MeetingAdapterError.invalidAudioChunk
    }
    self.schemaVersion = meetingAudioChunkSchemaVersion
    self.chunkID = chunkID
    self.sessionID = sessionID
    self.track = track
    self.sequence = sequence
    self.startMonotonicMilliseconds = startMonotonicMilliseconds
    self.endMonotonicMilliseconds = endMonotonicMilliseconds
    self.checksum = checksum
    self.byteLength = byteLength
    self.sealed = true
    self.codec = codec
    self.sampleRateHertz = sampleRateHertz
    self.channelCount = channelCount
  }

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case chunkID = "chunk_id"
    case sessionID = "session_id"
    case track
    case sequence
    case startMonotonicMilliseconds = "start_monotonic_ms"
    case endMonotonicMilliseconds = "end_monotonic_ms"
    case checksum
    case byteLength = "byte_length"
    case sealed
    case codec
    case sampleRateHertz = "sample_rate_hz"
    case channelCount = "channel_count"
  }
}

public struct CapturedAudioChunk: Equatable, Sendable {
  public let metadata: MeetingAudioChunk
  public let bytes: Data

  public init(metadata: MeetingAudioChunk, bytes: Data) throws {
    guard metadata.byteLength == bytes.count else {
      throw MeetingAdapterError.audioChunkLengthMismatch
    }
    self.metadata = metadata
    self.bytes = bytes
  }
}

public enum MeetingRealtimeAudioDerivation: String, Codable, Equatable, Sendable {
  case raw
  case appleVoiceProcessing = "apple_voice_processing"
}

public struct MeetingRealtimeAudioFrame: Codable, Equatable, Sendable {
  public let schemaVersion: String
  public let frameID: String
  public let sessionID: String
  public let track: MeetingAudioTrack
  public let frameSequence: Int
  public let sourceChunkID: String
  public let startMonotonicMilliseconds: Int64
  public let endMonotonicMilliseconds: Int64
  public let codec: String
  public let sampleRateHertz: Int
  public let channelCount: Int
  public let speechPresent: Bool?
  public let audioDerivation: MeetingRealtimeAudioDerivation?

  public init(
    frameID: String,
    sessionID: String,
    track: MeetingAudioTrack,
    frameSequence: Int,
    sourceChunkID: String,
    startMonotonicMilliseconds: Int64,
    endMonotonicMilliseconds: Int64,
    codec: String = "pcm_s16le",
    sampleRateHertz: Int = 16_000,
    channelCount: Int = 1,
    speechPresent: Bool? = nil,
    audioDerivation: MeetingRealtimeAudioDerivation? = nil
  ) throws {
    guard !frameID.isEmpty, !sessionID.isEmpty, frameSequence >= 0,
      !sourceChunkID.isEmpty, startMonotonicMilliseconds >= 0,
      endMonotonicMilliseconds > startMonotonicMilliseconds,
      codec == "pcm_s16le", sampleRateHertz == 16_000, channelCount == 1
    else { throw MeetingAdapterError.invalidAudioChunk }
    schemaVersion = meetingRealtimeAudioFrameSchemaVersion
    self.frameID = frameID
    self.sessionID = sessionID
    self.track = track
    self.frameSequence = frameSequence
    self.sourceChunkID = sourceChunkID
    self.startMonotonicMilliseconds = startMonotonicMilliseconds
    self.endMonotonicMilliseconds = endMonotonicMilliseconds
    self.codec = codec
    self.sampleRateHertz = sampleRateHertz
    self.channelCount = channelCount
    self.speechPresent = speechPresent
    self.audioDerivation = audioDerivation
  }

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case frameID = "frame_id"
    case sessionID = "session_id"
    case track
    case frameSequence = "frame_sequence"
    case sourceChunkID = "source_chunk_id"
    case startMonotonicMilliseconds = "start_monotonic_ms"
    case endMonotonicMilliseconds = "end_monotonic_ms"
    case codec
    case sampleRateHertz = "sample_rate_hz"
    case channelCount = "channel_count"
    case speechPresent = "speech_present"
    case audioDerivation = "audio_derivation"
  }
}

public struct CapturedRealtimeAudioFrame: Equatable, Sendable {
  public let metadata: MeetingRealtimeAudioFrame
  public let bytes: Data

  public init(metadata: MeetingRealtimeAudioFrame, bytes: Data) {
    self.metadata = metadata
    self.bytes = bytes
  }
}

public struct MeetingChunkReference: Codable, Equatable, Sendable {
  public let chunkID: String
  public let track: MeetingAudioTrack
  public let sequence: Int
  public let checksum: String

  enum CodingKeys: String, CodingKey {
    case chunkID = "chunk_id"
    case track
    case sequence
    case checksum
  }
}

public enum MeetingSessionAuditEventType: String, Codable, Sendable {
  case sessionDetected = "session.detected"
  case recordingStarted = "recording.started"
  case recordingPaused = "recording.paused"
  case recordingResumed = "recording.resumed"
  case audioChunkSealed = "audio.chunk.sealed"
  case audioTrackUnavailable = "audio.track.unavailable"
  case meetingEndConfirmed = "meeting.end.confirmed"
  case interruptedSessionRecovered = "session.interrupted.recovered"
  case recordingStopped = "recording.stopped"
}

public struct MeetingSessionAuditEvent: Codable, Equatable, Sendable {
  public let eventID: String
  public let type: MeetingSessionAuditEventType
  public let atMonotonicMilliseconds: Int64
  public let evidence: ConfirmedMeetingEndEvidence?
  public let chunkReference: MeetingChunkReference?
  public let stopReason: MeetingStopReason?
  public let track: MeetingAudioTrack?
  public let unavailabilityReason: String?

  public init(
    eventID: String,
    type: MeetingSessionAuditEventType,
    atMonotonicMilliseconds: Int64,
    evidence: ConfirmedMeetingEndEvidence?,
    chunkReference: MeetingChunkReference?,
    stopReason: MeetingStopReason?,
    track: MeetingAudioTrack? = nil,
    unavailabilityReason: String? = nil
  ) {
    self.eventID = eventID
    self.type = type
    self.atMonotonicMilliseconds = atMonotonicMilliseconds
    self.evidence = evidence
    self.chunkReference = chunkReference
    self.stopReason = stopReason
    self.track = track
    self.unavailabilityReason = unavailabilityReason
  }

  enum CodingKeys: String, CodingKey {
    case eventID = "event_id"
    case type
    case atMonotonicMilliseconds = "at_monotonic_ms"
    case evidence
    case chunkReference = "chunk_ref"
    case stopReason = "stop_reason"
    case track
    case unavailabilityReason = "unavailability_reason"
  }
}

public struct MeetingSessionState: Codable, Equatable, Sendable {
  public let schemaVersion: String
  public let sessionID: String
  public let platform: MeetingPlatform
  public let meetingReference: String
  public let startMode: MeetingStartMode
  public var status: MeetingSessionStatus
  public let wallClockAnchorMilliseconds: Int64
  public let monotonicAnchorMilliseconds: Int64
  public var startedMonotonicMilliseconds: Int64?
  public var endedMonotonicMilliseconds: Int64?
  public var stopReason: MeetingStopReason?
  public var events: [MeetingSessionAuditEvent]

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case sessionID = "session_id"
    case platform
    case meetingReference = "meeting_ref"
    case startMode = "start_mode"
    case status
    case wallClockAnchorMilliseconds = "wall_clock_anchor_ms"
    case monotonicAnchorMilliseconds = "monotonic_anchor_ms"
    case startedMonotonicMilliseconds = "started_monotonic_ms"
    case endedMonotonicMilliseconds = "ended_monotonic_ms"
    case stopReason = "stop_reason"
    case events
  }
}

public struct SealedMeetingSequenceManifest: Codable, Equatable, Sendable {
  public let schemaVersion = "inkloop.meeting_sequence_manifest.v1"
  public let sessionID: String
  public let expectedTracks: [MeetingAudioTrack]
  public let expectedLastSequence: [MeetingAudioTrack: Int]
  public let knownMissingChunkIDs: [String]

  public init(sealedSession session: MeetingSessionState) throws {
    guard session.status == .sealed else { throw MeetingAdapterError.sessionNotSealed }
    var lastSequence: [MeetingAudioTrack: Int] = [:]
    for event in session.events {
      guard let chunk = event.chunkReference else { continue }
      lastSequence[chunk.track] = max(lastSequence[chunk.track] ?? -1, chunk.sequence)
    }
    sessionID = session.sessionID
    // A completely silent or failed track is still part of the expected
    // evidence contract. Keeping both tracks here lets the server represent
    // that condition as `missing_track:*` instead of silently upgrading a
    // degraded capture to a complete transcript.
    expectedTracks = MeetingAudioTrack.allCases
    expectedLastSequence = lastSequence
    knownMissingChunkIDs = session.events.compactMap { event in
      guard event.type == .audioTrackUnavailable, let track = event.track else { return nil }
      return "track_unavailable:\(track.rawValue):\(event.atMonotonicMilliseconds)"
    }
  }

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case sessionID = "session_id"
    case expectedTracks = "expected_tracks"
    case expectedLastSequence = "expected_last_sequence"
    case knownMissingChunkIDs = "known_missing_chunk_ids"
  }

  public init(from decoder: any Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    sessionID = try container.decode(String.self, forKey: .sessionID)
    expectedTracks = try container.decode([MeetingAudioTrack].self, forKey: .expectedTracks)
    knownMissingChunkIDs = try container.decodeIfPresent(
      [String].self, forKey: .knownMissingChunkIDs) ?? []
    let encodedLastSequence = try container.decode(
      [String: Int].self, forKey: .expectedLastSequence)
    var decodedLastSequence: [MeetingAudioTrack: Int] = [:]
    for (rawTrack, sequence) in encodedLastSequence {
      guard let track = MeetingAudioTrack(rawValue: rawTrack) else {
        throw DecodingError.dataCorruptedError(
          forKey: .expectedLastSequence,
          in: container,
          debugDescription: "Unsupported meeting audio track: \(rawTrack)"
        )
      }
      decodedLastSequence[track] = sequence
    }
    expectedLastSequence = decodedLastSequence
  }

  public func encode(to encoder: any Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(schemaVersion, forKey: .schemaVersion)
    try container.encode(sessionID, forKey: .sessionID)
    try container.encode(expectedTracks, forKey: .expectedTracks)
    try container.encode(
      Dictionary(uniqueKeysWithValues: expectedLastSequence.map { ($0.key.rawValue, $0.value) }),
      forKey: .expectedLastSequence
    )
    try container.encode(knownMissingChunkIDs, forKey: .knownMissingChunkIDs)
  }
}

public enum MeetingAdapterError: Error, Equatable, Sendable {
  case alreadyRecording
  case noActiveRecording
  case autoRecordDisabled
  case invalidConfirmedEndEvidence
  case weakMeetingEndSignal(String)
  case invalidAudioChunk
  case audioChunkLengthMismatch
  case chunkBelongsToAnotherSession
  case chunkConflict(String)
  case confirmedEndBelongsToAnotherMeeting
  case captureTransitionInProgress
  case noAudioTracksAvailable
  case sessionNotSealed
  case invalidStorageIdentifier(String)
  case capturePermissionDenied(String)
  case targetApplicationUnavailable(MeetingPlatform)
  case unsupportedAudioFormat
  case captureNotPaused
  case captureAlreadyPaused
  case invalidAcknowledgement
  case recorderLeaseHeldByAnotherDevice(String)
}

File: packages/meeting-media-core/src/contracts.ts
export const MEETING_SESSION_SCHEMA_VERSION = 'inkloop.meeting_session.v1' as const;
export const MEETING_AUDIO_CHUNK_SCHEMA_VERSION = 'inkloop.meeting_audio_chunk.v1' as const;
export const MEETING_REALTIME_AUDIO_FRAME_SCHEMA_VERSION = 'inkloop.meeting_realtime_audio_frame.v1' as const;
export const MEETING_CHUNK_DELIVERY_SCHEMA_VERSION = 'inkloop.meeting_chunk_delivery.v1' as const;
export const MEETING_TRANSCRIPT_SCHEMA_VERSION = 'inkloop.meeting_transcript.v1' as const;

export type MeetingPlatform = 'google_meet' | 'zoom';
export type MeetingSessionStartMode = 'automatic' | 'manual';
export type MeetingSessionStatus = 'detected' | 'recording' | 'paused' | 'sealed';
export type MeetingAudioTrack = 'mic' | 'remote';
export type MeetingRealtimeAudioDerivation = 'raw' | 'apple_voice_processing';

export interface MeetingContractValidationIssue {
  path: string;
  message: string;
}

export interface ConfirmedMeetingEndEvidence {
  adapter: string;
  signal: string;
  provider_event_id?: string;
  observed_at_wall_clock_ms?: number;
}

export interface MeetingAudioChunk {
  schema_version: typeof MEETING_AUDIO_CHUNK_SCHEMA_VERSION;
  chunk_id: string;
  session_id: string;
  track: MeetingAudioTrack;
  sequence: number;
  start_monotonic_ms: number;
  end_monotonic_ms: number;
  checksum: string;
  byte_length: number;
  sealed: boolean;
  codec?: string;
  sample_rate_hz?: number;
  channel_count?: number;
}

/**
 * Ephemeral low-latency ASR projection input. The enclosing 5-second
 * MeetingAudioChunk remains the immutable fact and retry boundary.
 */
export interface MeetingRealtimeAudioFrame {
  schema_version: typeof MEETING_REALTIME_AUDIO_FRAME_SCHEMA_VERSION;
  frame_id: string;
  session_id: string;
  track: MeetingAudioTrack;
  frame_sequence: number;
  source_chunk_id: string;
  start_monotonic_ms: number;
  end_monotonic_ms: number;
  codec: 'pcm_s16le';
  sample_rate_hz: number;
  channel_count: 1;
  /**
   * Client-side speech projection decision when one was made. A false frame
   * carries no audio and advances transport coverage. An omitted value means
   * the server must apply its own VAD (the default for Remote).
   */
  speech_present?: boolean;
  /**
   * Identifies the ephemeral ASR projection. Authoritative fact chunks remain
   * raw even when this frame was produced by Apple Voice Processing.
   */
  audio_derivation?: MeetingRealtimeAudioDerivation;
}

export interface MeetingAudioChunkAcknowledgement {
  session_id: string;
  track: MeetingAudioTrack;
  sequence: number;
  chunk_id: string;
  checksum: string;
  acknowledged_at_ms: number;
}

export interface MeetingChunkDeliveryState {
  schema_version: typeof MEETING_CHUNK_DELIVERY_SCHEMA_VERSION;
  session_id: string;
  chunks: Record<string, MeetingAudioChunk>;
  acknowledgements: Record<string, MeetingAudioChunkAcknowledgement>;
}

export interface MeetingUtterance {
  utterance_id: string;
  session_id: string;
  track: MeetingAudioTrack;
  start_ms: number;
  end_ms: number;
  text: string;
  revision: number;
  stability: 'provisional' | 'formal';
  source_chunk_ids: string[];
  speaker_cluster_id?: string;
  confidence?: number;
  duplicate_of?: string;
}

export interface MeetingTranscriptState {
  schema_version: typeof MEETING_TRANSCRIPT_SCHEMA_VERSION;
  session_id: string;
  status: 'provisional' | 'formal';
  finality?: 'final' | 'partial';
  utterances: MeetingUtterance[];
  missing_chunks: string[];
  revision: number;
  finalized_at_ms?: number;
}

export interface MeetingSessionAuditEvent {
  event_id: string;
  type:
    | 'session.detected'
    | 'recording.started'
    | 'recording.paused'
    | 'recording.resumed'
    | 'audio.chunk.sealed'
    | 'audio.track.unavailable'
    | 'meeting.end.confirmed'
    | 'session.interrupted.recovered'
    | 'recording.stopped';
  at_monotonic_ms: number;
  evidence?: ConfirmedMeetingEndEvidence;
  chunk_ref?: Pick<MeetingAudioChunk, 'chunk_id' | 'track' | 'sequence' | 'checksum'>;
  stop_reason?: 'meeting_end_confirmed' | 'manual' | 'interrupted_session_recovered';
  track?: MeetingAudioTrack;
  unavailability_reason?: string;
}

export interface MeetingSessionState {
  schema_version: typeof MEETING_SESSION_SCHEMA_VERSION;
  session_id: string;
  platform: MeetingPlatform;
  meeting_ref: string;
  start_mode: MeetingSessionStartMode;
  status: MeetingSessionStatus;
  wall_clock_anchor_ms: number;
  monotonic_anchor_ms: number;
  started_monotonic_ms?: number;
  ended_monotonic_ms?: number;
  stop_reason?: 'meeting_end_confirmed' | 'manual' | 'interrupted_session_recovered';
  events: MeetingSessionAuditEvent[];
}

export type MeetingSessionEvent =
  | { type: 'recording.started'; at_monotonic_ms: number }
  | { type: 'recording.paused'; at_monotonic_ms: number }
  | { type: 'recording.resumed'; at_monotonic_ms: number }
  | { type: 'audio.chunk.sealed'; at_monotonic_ms: number; chunk: MeetingAudioChunk }
  | { type: 'audio.track.unavailable'; at_monotonic_ms: number; track: MeetingAudioTrack; reason: string }
  | { type: 'meeting.end.confirmed'; at_monotonic_ms: number; evidence: ConfirmedMeetingEndEvidence }
  | { type: 'session.interrupted.recovered'; at_monotonic_ms: number }
  | { type: 'recording.stopped'; at_monotonic_ms: number; reason: 'manual' };

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

function requireNonEmptyString(
  value: Record<string, unknown>,
  key: string,
  issues: MeetingContractValidationIssue[],
): void {
  if (typeof value[key] !== 'string' || value[key] === '') issues.push({ path: key, message: 'must be a non-empty string' });
}

function requireFiniteNumber(
  value: Record<string, unknown>,
  key: string,
  issues: MeetingContractValidationIssue[],
): void {
  if (typeof value[key] !== 'number' || !Number.isFinite(value[key])) issues.push({ path: key, message: 'must be a finite number' });
}

export function validateMeetingAudioChunk(value: unknown): MeetingContractValidationIssue[] {
  if (!isRecord(value)) return [{ path: '', message: 'must be an object' }];
  const issues: MeetingContractValidationIssue[] = [];
  if (value.schema_version !== MEETING_AUDIO_CHUNK_SCHEMA_VERSION) {
    issues.push({ path: 'schema_version', message: `must be ${MEETING_AUDIO_CHUNK_SCHEMA_VERSION}` });
  }
  requireNonEmptyString(value, 'chunk_id', issues);
  requireNonEmptyString(value, 'session_id', issues);
  requireNonEmptyString(value, 'checksum', issues);
  if (value.track !== 'mic' && value.track !== 'remote') issues.push({ path: 'track', message: 'must be mic or remote' });
  if (!Number.isInteger(value.sequence) || Number(value.sequence) < 0) {
    issues.push({ path: 'sequence', message: 'must be a non-negative integer' });
  }
  requireFiniteNumber(value, 'start_monotonic_ms', issues);
  requireFiniteNumber(value, 'end_monotonic_ms', issues);
  if (typeof value.start_monotonic_ms === 'number' && typeof value.end_monotonic_ms === 'number' && value.end_monotonic_ms <= value.start_monotonic_ms) {
    issues.push({ path: 'end_monotonic_ms', message: 'must be greater than start_monotonic_ms' });
  }
  if (!Number.isInteger(value.byte_length) || Number(value.byte_length) < 0) {
    issues.push({ path: 'byte_length', message: 'must be a non-negative integer' });
  }
  if (value.sealed !== true) issues.push({ path: 'sealed', message: 'must be true before delivery' });
  return issues;
}

export function assertMeetingAudioChunk(value: unknown): asserts value is MeetingAudioChunk {
  const issues = validateMeetingAudioChunk(value);
  if (issues.length > 0) throw new Error(`invalid meeting audio chunk: ${issues.map((issue) => `${issue.path} ${issue.message}`).join(', ')}`);
}

export function validateMeetingRealtimeAudioFrame(value: unknown): MeetingContractValidationIssue[] {
  if (!isRecord(value)) return [{ path: '', message: 'must be an object' }];
  const issues: MeetingContractValidationIssue[] = [];
  if (value.schema_version !== MEETING_REALTIME_AUDIO_FRAME_SCHEMA_VERSION) {
    issues.push({ path: 'schema_version', message: `must be ${MEETING_REALTIME_AUDIO_FRAME_SCHEMA_VERSION}` });
  }
  requireNonEmptyString(value, 'frame_id', issues);
  requireNonEmptyString(value, 'session_id', issues);
  requireNonEmptyString(value, 'source_chunk_id', issues);
  if (value.track !== 'mic' && value.track !== 'remote') issues.push({ path: 'track', message: 'must be mic or remote' });
  if (!Number.isInteger(value.frame_sequence) || Number(value.frame_sequence) < 0) {
    issues.push({ path: 'frame_sequence', message: 'must be a non-negative integer' });
  }
  requireFiniteNumber(value, 'start_monotonic_ms', issues);
  requireFiniteNumber(value, 'end_monotonic_ms', issues);
  if (typeof value.start_monotonic_ms === 'number' && typeof value.end_monotonic_ms === 'number'
    && value.end_monotonic_ms <= value.start_monotonic_ms) {
    issues.push({ path: 'end_monotonic_ms', message: 'must be greater than start_monotonic_ms' });
  }
  if (value.codec !== 'pcm_s16le') issues.push({ path: 'codec', message: 'must be pcm_s16le' });
  if (value.sample_rate_hz !== 16_000) issues.push({ path: 'sample_rate_hz', message: 'must be 16000' });
  if (value.channel_count !== 1) issues.push({ path: 'channel_count', message: 'must be 1' });
  if (value.speech_present !== undefined && typeof value.speech_present !== 'boolean') {
    issues.push({ path: 'speech_present', message: 'must be a boolean when provided' });
  }
  if (value.audio_derivation !== undefined
    && value.audio_derivation !== 'raw'
    && value.audio_derivation !== 'apple_voice_processing') {
    issues.push({ path: 'audio_derivation', message: 'must be raw or apple_voice_processing when provided' });
  }
  return issues;
}

export function assertMeetingRealtimeAudioFrame(value: unknown): asserts value is MeetingRealtimeAudioFrame {
  const issues = validateMeetingRealtimeAudioFrame(value);
  if (issues.length > 0) {
    throw new Error(`invalid meeting realtime audio frame: ${issues.map((issue) => `${issue.path} ${issue.message}`).join(', ')}`);
  }
}
